content/blog/package-react-components/index.md

```diff
@@ -0,0 +1,776 @@
+and pieces of code in general, between apps usually inside a private NPM repository, but
+it also apply to Open Source software like a single Component package such as a Date Picker
+or a Multi Select and also to Component Libraries or UI Frameworks like [React Bootstrap][9],
+[Ant Design][8], [Semantic UI][4], [Material UI][7], etc (more on these later).
+
+This problem might seem simple enough but it turns out that it has a lot of nuances and trade-offs.
+
+The main challenge in packaging a [React][1] Component is surprisingly not Javascript,
+but all the Assets that might be related to that component such as CSS, Images, SVGs and fonts.
+These assets are directly related to our Javascript code via some of the following statements:
+
+- `imports image from './image.{png,jpg,svg}'` in Javascript
+- `import classNames from './styles.{css,scss,less)'` in Javascript
+- `url(image.{png,jpg,svg})` in CSS
+- `font-family and @import` in CSS
+
+[Webpack][2] can understand all these dependency statements via _loaders_ and completely
+analyse the cross-language and cross-file-type dependency graph, which for a very simple
+component might look something like this.
+
+```
+            --> styles.css --> otherImage.svg
+MyComponent --> image.svg
+
+```
+
+It turns out that there are a lot of schemes for representing this graph in a Packaged NPM module,
+each with it's own trade-offs. Analysing each alternative is the main purpose of this post.
+
+Unfortunately in early 2020 there still no universal standard way of Packaging a Front End module with all
+its assets, but a lot of progress has been made and there are actually lots of promising solutions.
+
+We are going to explore how this problem is tackled in the wild by studying
+each technique and providing example real world production ready libraries that
+use them and in the end of this post we are going
+to recommend a particular solution to this problem based of my experience.
+
+Additionally we are going to explore how each of these packaging schemes enable
+different **Theming** schemes and, at the end, recommend one.
+
+Most of this post is also applicable to any piece of Front End Javascript code that is related
+to CSS, images, etc, such as [Angular][3] Components / Directives, [Vue][15] Components, etc.
+
+## To Inline or not to Inline
+
+In the context of Front End assets, inlining means to bake those assets directly
+into the Javascript code (bundle) generated by our bundlers or compilers such as Webpack or [Babel][4].
+
+CSS can be inlined without any third party library but to enable all regular CSS features
+a third party **CSS-in-JS** library is recommended such as [Emotion][5] or [Styled-components][6], more on this later.
+Images can be inlined by Webpack's [url-loader][11] in the form of [Data URL][100] with base 64 encoding.
+SVG can be inlined by turning an `.svg` file into a [React][1] Component (An SVG file is just a regular HTML tree)
```

Comment:
Unless I misunderstood, the SVG->React transformation means to manually create a component that returns the SVG tag for a specific SVG, right? I know what a React Component looks like and I know what an SVG looks like, but I'm not 100% sure that what I assumed is correct.

Reply:
yes, that is right

Reply:
something like this

```javascript
export default function MyComp() {
  return <svg><circle r=10 /></svg>
}
```

This will render an svg circle (if I actually wrote this correctly) but the main problem with this is that you need to transform regular svg files into this format, most of the time you cannot simply copy the contents of an svg file into the return statement of a react componet because jsx is html and not xml so things such as xml namespaces and other anotations that regular svg drawing tools put into the svg xml tree need to be manually removed.

Being all that said, this is a really great way of dealing with svg because you can modify the svg at runtime via props and assigning those props to different svg attributes such as paths, sizes, colors etc.

Reply:
maybe I can add this to the post? not sure where, maybe an appendix?

Reply:
I think it depends on the target, I guess. Your introcution helped me but maybe something who has at least some experience with React does not need it.

Reply:
Added a link to a post that explores this. In fact there is a lib that automatically turns svg into react